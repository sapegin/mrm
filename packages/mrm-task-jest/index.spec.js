jest.mock('fs');
jest.mock('mrm-core/src/util/log', () => ({
	added: jest.fn(),
}));
jest.mock('mrm-core/src/npm', () => ({
	install: jest.fn(),
	uninstall: jest.fn(),
}));

const fs = jest.requireActual('fs');
const path = require('path');
const { install, uninstall } = require('mrm-core');
const vol = require('memfs').vol;
const task = require('./index');

const console$log = console.log;

const stringify = o => JSON.stringify(o, null, '  ');

const packageJson = stringify({
	name: 'unicorn',
});

beforeEach(() => {
	console.log = jest.fn();
});

afterEach(() => {
	vol.reset();
	install.mockClear();
	uninstall.mockClear();
	console.log = console$log;
});

it('should add Jest', () => {
	vol.fromJSON({
		'/package.json': packageJson,
	});

	task({});

	expect(vol.toJSON()).toMatchSnapshot();
	expect(install).toBeCalledWith(['jest']);
});

it('should add Babel specific configuration if projects depends on Babel', () => {
	vol.fromJSON({
		'/package.json': stringify({
			name: 'unicorn',
			devDependencies: {
				'babel-core': '*',
			},
		}),
	});

	task({});

	expect(vol.toJSON()).toMatchSnapshot();
	expect(install).toBeCalledWith(['jest', 'babel-jest']);
});

it('should add TypeScript specific configuration if projects depends on TypeScript', () => {
	vol.fromJSON({
		'/package.json': stringify({
			name: 'unicorn',
			devDependencies: {
				typescript: '*',
			},
		}),
	});

	task({});

	expect(vol.toJSON()).toMatchSnapshot();
	expect(install).toBeCalledWith(['jest', 'ts-jest', '@types/jest']);
});

it('should add React specific configuration if projects depends on React', () => {
	vol.fromJSON({
		[`${__dirname}/templates/jestsetup.js`]: fs
			.readFileSync(path.join(__dirname, 'templates/jestsetup.js'))
			.toString(),
		'/package.json': stringify({
			name: 'unicorn',
			dependencies: {
				react: '*',
			},
		}),
	});
	task({});

	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
	expect(vol.toJSON()['/test/jestsetup.js']).toMatchSnapshot();
	expect(install).toBeCalledWith(expect.arrayContaining(['jest', 'enzyme']));
});

it('should not overwrite Jest setup file', () => {
	vol.fromJSON({
		[`${__dirname}/templates/jestsetup.js`]: fs
			.readFileSync(path.join(__dirname, 'templates/jestsetup.js'))
			.toString(),
		'/test/jestsetup.js': 'pizza',
		'/package.json': stringify({
			name: 'unicorn',
			dependencies: {
				react: '*',
			},
		}),
	});

	task({});

	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
	expect(vol.toJSON()['/test/jestsetup.js']).toBe('pizza');
});

it('should not overwrite Jest setup file (with enzyme)', () => {
	vol.fromJSON({
		[`${__dirname}/templates/jestsetup.js`]: fs
			.readFileSync(path.join(__dirname, 'templates/jestsetup.js'))
			.toString(),
		'/test/jestsetup.js': "import Adapter from 'enzyme-adapter-react-16';",
		'/package.json': stringify({
			name: 'unicorn',
			dependencies: {
				react: '*',
			},
		}),
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
	expect(vol.toJSON()['/test/jestsetup.js']).toBe(
		"import Adapter from 'enzyme-adapter-react-16';"
	);
});

it('should update or create .eslintignore if projects depends on ESLint', () => {
	vol.fromJSON({
		'/package.json': stringify({
			name: 'unicorn',
			devDependencies: {
				eslint: '*',
				'babel-core': '*',
			},
		}),
	});

	task({});

	expect(vol.toJSON()['/.eslintignore']).toMatchSnapshot();
});

it('should update or create .eslintignore if projects depends on ESLint (no babel-core)', () => {
	vol.fromJSON({
		'/package.json': stringify({
			name: 'unicorn',
			devDependencies: {
				eslint: '*',
			},
		}),
	});

	task(getConfigGetter({}));

	expect(vol.toJSON()['/.eslintignore']).toMatchSnapshot();
});

it('should add a basic test case when index.js file is present', () => {
	vol.fromJSON({
		[`${__dirname}/templates/test.js`]: fs
			.readFileSync(path.join(__dirname, 'templates/test.js'))
			.toString(),
		'/package.json': stringify({
			name: 'uber-unicorn',
		}),
		'/index.js': '',
	});

	task({});

	expect(vol.toJSON()['/test.js']).toMatchSnapshot();
});

it('should not update .npmignore for private packages', () => {
	vol.fromJSON({
		'/package.json': stringify({
			name: 'unicorn',
			private: true,
		}),
	});

	task({});

	expect(vol.toJSON()['/.npmignore']).toBeUndefined();
});

it('should not overwrite test.js', () => {
	vol.fromJSON({
		'/package.json': packageJson,
		'/index.js': '',
		'/test.js': 'still here',
	});

	task({});

	expect(vol.toJSON()['/test.js']).toMatchSnapshot();
});

it('should uninstall other test frameworks', () => {
	vol.fromJSON({
		'/package.json': stringify({
			name: 'unicorn',
			devDependencies: {
				mocha: '*',
			},
		}),
	});

	task({});

	expect(uninstall).toBeCalledWith(['mocha', 'chai', 'ava']);
	expect(console.log).toBeCalledWith(
		expect.stringMatching('Migrate your tests to Jest')
	);
});
