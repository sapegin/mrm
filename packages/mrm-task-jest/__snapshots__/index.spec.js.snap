// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add Babel specific configuration if projects depends on Babel 1`] = `
Object {
  "/.gitignore": "coverage/",
  "/.npmignore": "__tests__/",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"devDependencies\\": {
    \\"babel-core\\": \\"*\\"
  },
  \\"scripts\\": {
    \\"test:jest\\": \\"jest\\",
    \\"test:watch\\": \\"jest --watch\\",
    \\"test:coverage\\": \\"jest --coverage\\",
    \\"test\\": \\"npm run test:jest --\\"
  },
  \\"jest\\": {
    \\"testPathIgnorePatterns\\": [
      \\"/node_modules/\\",
      \\"<rootDir>/lib/\\"
    ]
  }
}",
}
`;

exports[`should add Jest 1`] = `
Object {
  "/.gitignore": "coverage/",
  "/.npmignore": "__tests__/",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {
    \\"test:jest\\": \\"jest\\",
    \\"test:watch\\": \\"jest --watch\\",
    \\"test:coverage\\": \\"jest --coverage\\",
    \\"test\\": \\"npm run test:jest --\\"
  }
}",
}
`;

exports[`should add React specific configuration if projects depends on React 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"dependencies\\": {
    \\"react\\": \\"*\\"
  },
  \\"scripts\\": {
    \\"test:jest\\": \\"jest\\",
    \\"test:watch\\": \\"jest --watch\\",
    \\"test:coverage\\": \\"jest --coverage\\",
    \\"test\\": \\"npm run test:jest --\\"
  },
  \\"jest\\": {
    \\"snapshotSerializers\\": [
      \\"enzyme-to-json/serializer\\"
    ],
    \\"setupFiles\\": [
      \\"./test/jestsetup.js\\"
    ]
  }
}"
`;

exports[`should add React specific configuration if projects depends on React 2`] = `
"import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });
"
`;

exports[`should add TypeScript specific configuration if projects depends on TypeScript 1`] = `
Object {
  "/.gitignore": "coverage/",
  "/.npmignore": "__tests__/",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"devDependencies\\": {
    \\"typescript\\": \\"*\\"
  },
  \\"scripts\\": {
    \\"test:jest\\": \\"jest\\",
    \\"test:watch\\": \\"jest --watch\\",
    \\"test:coverage\\": \\"jest --coverage\\",
    \\"test\\": \\"npm run test:jest --\\"
  },
  \\"jest\\": {
    \\"testPathIgnorePatterns\\": [
      \\"/node_modules/\\",
      \\"<rootDir>/lib/\\"
    ],
    \\"transform\\": {
      \\"^.+\\\\\\\\.tsx?$\\": \\"<rootDir>/node_modules/ts-jest/preprocessor.js\\"
    },
    \\"testRegex\\": \\"(/__tests__/.*|(\\\\\\\\.|/)(test|spec))\\\\\\\\.(jsx?|tsx?)$\\",
    \\"moduleFileExtensions\\": [
      \\"ts\\",
      \\"tsx\\",
      \\"js\\",
      \\"jsx\\",
      \\"json\\"
    ]
  }
}",
}
`;

exports[`should add a basic test case when index.js file is present 1`] = `
"const uberUnicorn = require('./index');

test('uberUnicorn', () => {
	const result = uberUnicorn();
	expect(result).toBeTruthy();
});
"
`;

exports[`should not overwrite Jest setup file 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"dependencies\\": {
    \\"react\\": \\"*\\"
  },
  \\"scripts\\": {
    \\"test:jest\\": \\"jest\\",
    \\"test:watch\\": \\"jest --watch\\",
    \\"test:coverage\\": \\"jest --coverage\\",
    \\"test\\": \\"npm run test:jest --\\"
  },
  \\"jest\\": {
    \\"snapshotSerializers\\": [
      \\"enzyme-to-json/serializer\\"
    ]
  }
}"
`;

exports[`should not overwrite Jest setup file (with enzyme) 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"dependencies\\": {
    \\"react\\": \\"*\\"
  },
  \\"scripts\\": {
    \\"test:jest\\": \\"jest\\",
    \\"test:watch\\": \\"jest --watch\\",
    \\"test:coverage\\": \\"jest --coverage\\",
    \\"test\\": \\"npm run test:jest --\\"
  },
  \\"jest\\": {
    \\"snapshotSerializers\\": [
      \\"enzyme-to-json/serializer\\"
    ]
  }
}"
`;

exports[`should not overwrite test.js 1`] = `"still here"`;

exports[`should update or create .eslintignore if projects depends on ESLint (no babel-core) 1`] = `"coverage/*"`;

exports[`should update or create .eslintignore if projects depends on ESLint 1`] = `
"coverage/*
lib/*"
`;
